packet_encoder.go

```diff
@@ -19,9 +19,15 @@ type packetEncoder interface {
 	putInt32Array(in []int32) error
 	putInt64Array(in []int64) error
 
+	// Provide the current offset to record the batch size metric
+	offset() int
+
 	// Stacks, see PushEncoder
 	push(in pushEncoder)
 	pop() error
+
+	// To only record metrics during the real encoder pass
+	doRecordMetrics() bool
```

Comment:
> (it's an interface)

Ah, missed that the first time round.

> Do you prefer to add a metrics.Registry argument to encode(e encoder) ([]byte, error) or add another function encodeWithMetrics(e encoder, metricsRegistry metrics.Registry) ([]byte, error) like in the initial 1e5e537 commit?

Just add an argument to `encode` and have the other callers pass `nil`, this doesn't feel important enough to warrant a whole method.